R/subscript-loc.R

```diff
@@ -62,70 +54,110 @@ vec_as_location <- function(i,
                             n,
                             names = NULL,
                             ...,
-                            allow_types = c("indicator", "location", "name"),
-                            convert_values = "negative",
                             arg = "i") {
   if (!missing(...)) ellipsis::check_dots_empty()
 
-  n <- vec_coercible_cast(n, integer())
-  vec_assert(n, integer(), 1L)
-  i <- vec_as_subscript(i, arg = arg, allow_types = allow_types)
+  i <- vec_as_subscript(i, arg = arg)
+  .Call(vctrs_as_location, i, n, names, "invert")
+}
+#' @rdname vec_as_location
+#' @param negative Whether to `"invert"` negative values to positive
+#'   locations, throw an informative `"error"`, or `"ignore"` them.
+#' @export
+num_as_location <- function(i,
+                            n,
+                            names = NULL,
+                            ...,
+                            negative = c("invert", "error", "ignore"),
+                            arg = "i") {
+  if (!missing(...)) ellipsis::check_dots_empty()
 
-  convert_values <- as_opts_location_convert_values(convert_values, arg = arg)
-  .Call(vctrs_as_location, i, n, names, convert_values)
+  if (!is_integer(i) && !is_double(i)) {
+    abort("`i` must be a numeric vector.")
+  }
+  .Call(vctrs_as_location, i, n, names, negative)
 }
+
 #' @rdname vec_as_location
-#' @param allow_values Experimental. Character vector indicating zero,
-#'   one or several types of values to be allowed as input:
-#'   `"negative"` or `"missing"`. By default, locations can't be
-#'   negative or missing.
+#' @param missing Whether to throw an `"error"` when `i` is a missing
+#'   value, or `"ignore"` it.
 #' @export
 vec_as_location2 <- function(i,
                              n,
                              names = NULL,
                              ...,
-                             allow_values = NULL,
+                             missing = c("error", "ignore"),
```

Comment:
What does ignoring a value mean in this context? i.e. what gets returned instead?

Reply:
The value is returned, so a missing value in this case. I can see how `"ignore"` can be confusing. Would `"allow"` be better? The challenge is to find a term that describes the base case in other validation functions, for consistency.

Reply:
That's fine; just documentation that it returns an `NA`